docker/tarfile/src.go

```diff
@@ -35,14 +38,55 @@ type layerInfo struct {
 	size int64
 }
 
-// NewSource returns a tarfile.Source for the specified path.
-func NewSource(path string) *Source {
+// NewSourceFromFile returns a tarfile.Source for the specified path
+func NewSourceFromFile(path string) (*Source, error) {
+	file, err := os.Open(path)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error opening file %q", path)
+	}
+	defer file.Close()
+
+	reader, err := gzip.NewReader(file)
+	if err != nil {
+		return &Source{
+			tarPath: path,
+		}, nil
+	}
+	defer reader.Close()
+
+	return NewSourceFromStream(reader)
+}
+
+// NewSourceFromStream returns a tarfile.Source for the specified inputStream.
+func NewSourceFromStream(inputStream io.ReadCloser) (*Source, error) {
 	// TODO: We could add support for multiple images in a single archive, so
```

Comment:
This `TODO` applies to both functions equally, and is now a bit out of place.  Maybe move it to the doc string for `Source` itself, or as a stand-alone comment near the top of the file.

Reply:
moved it above the two source functions